### DISCUSSION TITLE
Debugging the firmware

### DISCUSSION THREAD

**Discussion by mmehari** [espotek-org/Labrador#276] [General]
Hi,

My question is more of an understanding to the internals of the project rather than finding and fixing a bug, but I would expect you used debugging tools during the time you build the firmware.

I would guess it will be through the PDI interface (PDI_CLK and PDI_DATA) already integrated in the espotek hardware but I don't know further details on its operation.

If you had the experience, would you be kind to share me the information?

**Comment by EspoTek** [comment]
Yep, your guess is bang on the money.

I debugged on Windows using Atmel Studio 7 and a PDI debugger.

In terms of hardware, I used the ATMEL JTAG-ICE here:
https://au.element14.com/microchip/atatmel-ice-pcba/debugger-atmel-arm-avr-pcba-kit/dp/2407171

I wouldn't recommend that exact bit of hardware, though.  You can get much cheaper PDI debuggers these days from AliExpress, Sparkfun etc.  and it looks like the price has tripled for the official kit since I bought mine!

In terms of how to debug in Atmel Studio, there's a little "Play button" at the top of the IDE that launches it in debugging mode.  You can set breakpoints just like you would in Visual Studio.  I can't remember exactly how to bring it up, but there's also a little window that lets you read the state of every single hardware register while execution is suspended.  Just make sure to set your PDI clock to the max or you're going to have a bad time!

**Comment by mi-hol** [comment]
@mmehari would this knowledge be worth capturing in wiki?

**Comment by mmehari** [comment]
Firmware debugging is indeed an important topic to capture in the Wiki. However, it is not that easy when it comes to PDI debuggers. They are not cheap, and I am not sure if there is an open source alternative.
Nevertheless, I would be happy to extend the [Firmware and Bootloader Mode](https://github.com/espotek-org/Labrador/wiki#firmware-and-bootloader-mode) section with firmware debugging.

**Comment by mi-hol** [comment]
> I would be happy to extend the [Firmware and Bootloader Mode](https://github.com/espotek-org/Labrador/wiki#firmware-and-bootloader-mode) section with firmware debugging.

Sounds great, please go ahead. let me know in case of any issues encountered

**Comment by mi-hol** [comment]
#260 might be another source of knowledge
